src/Integrations/Integrations/SpanTaxonomy.php

```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace DDTrace\Integrations;
+
+use DDTrace\SpanData;
+
+final class SpanTaxonomy
+{
+    /** @var string|null */
+    private static $currentRootService = null;
+
+    /**
+     * @param SpanData $span
+     * @param string $fallbackName
+     */
+    public static function handleInternalSpanServiceName(SpanData $span, $fallbackName = null)
+    {
+        $flatServiceNames =
+            \PHP_MAJOR_VERSION > 5
+                && \dd_trace_env_config('DD_TRACE_REMOVE_INTEGRATION_SERVICE_NAMES_ENABLED');
+
+        if ($flatServiceNames) {
+            $span->service = empty(self::$currentRootService)
+                ? \ddtrace_config_app_name($fallbackName)
+                : self::$currentRootService;
+        } else {
+            $span->service = empty($fallbackName) ? \ddtrace_config_app_name() : $fallbackName;
+        }
```

Comment:
I think this would be the better logic:
```suggestion
        if (!$flatServiceNames && !empty($fallbackName)) {
            $span->service = $fallbackName;
        }
```
The reason being: service names are automatically inherited on span creation. If nothing happens, it will be simply the service of the parent span, i.e. we remain in context of the actually executing service.

For example, you have a non-trivial application which is forked or using fibers, where you want to assign different forks/fibers different services. With the current approach, the DD_SERVICE would appear in the leaf functions, instead of being part of their actual parent service where it logically pertains to.

With the logic simplified that much, registerCurrentRootService will be unnecessary and this a simple helper function which could live in `Integration.php`.

Reply:
Applying some changes as discussed in person